### ISSUE TITLE
Commitlog header

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#1667]
Commitlog segments should include a header with the following info:

 - format information (for version upgrades)
 - schema (to simplify schema management -- dump schema once into a commitlog segment, close segment if schema changes)
 - shard information
 - probably more

the header should be extensible.

**Comment by tgrabiec** [comment]
If we put schema in the header that means we need to dump schema of all tables whenever a segment is opened because we don't know which ones will be written. Currently we append schema information to the segment when table is first used in the segment. This scales better as the number of tables grows.

**Comment by avikivity** [comment]
True.  When there are many tables, most are likely to share schemas, so we can take advantage of that.

**Comment by elcallio** [comment]
commit log segments encode shard + version in file name.

**Comment by avikivity** [comment]
Yes, but they can't encode any other information there.

Version numbers are a bad way of expressing features (can't selectively enable features), and encoding information in filenames is fragile.

For example, if we enable commitlog compression, we'll add a flag in the header to say so, and specify the algorithm. Ditto for at-rest encryption.

**Comment by elcallio** [comment]
Indeed. I agree a commit log format supporting compression or similar needs to have a header with more info. (with optional stuff in it). The header format of course indicated by a bumped format version. :-) (and I guess a magic @ start)

Just saying that the stuff you explicitly asked for is already in the segment (in one way or another). 

Compression otoh might need to be enabled per mutation even? Overhead might make it bad for very small mutations for example...

**Comment by avikivity** [comment]
For the cache and new sstable format, I want to selectively compress large blobs (and not touch metadata and small data).  For commitlog, since it is sequential access only, wholesale compression may make more sense.

**Comment by elcallio** [comment]
Compressing the metadata in commit log segments might not be good, because we want as much as possible of the file to be recoverable even in the event of disk errors. Thus, having the "chunked" format which allows easily finding next, hopefully undamaged, section in file is probably still desirable. One could compress per-chunk of course.
